crates/wasmtime/src/runtime/gc/enabled/arrayref.rs

```diff
@@ -385,14 +498,121 @@ impl ArrayRef {
             allocator.store_id,
             "attempted to use a `ArrayRefPre` with the wrong store"
         );
+        assert!(
+            !store.async_support(),
+            "use `ArrayRef::new_fixed_async` with asynchronous stores"
+        );
+
+        // Type check the elements against the element type.
+        for elem in elems {
+            elem.ensure_matches_ty(store, allocator.ty.element_type().unpack())
+                .context("element type mismatch")?;
+        }
```

Comment:
Could this get sunk into `_new_unchecked`? (making it a bit less unchecked)

Basically trying to get these multiple entrypoints as a one-liner-ish around the "guts" and the same "guts" are used everywhere